asmcomp/cmmgen.ml

```diff
@@ -1982,6 +1982,17 @@ and transl_prim_1 env p arg dbg =
                    [untag_int (transl env arg) dbg],
                    dbg))
               dbg
+  | Pcoerce coerce_to ->
```

Comment:
Coerce_to_int might not be strictly required, but I think it is cleaner to have it also.

Yes I think this is correct and the suggested changes should fix it.
As for cmmgen optimizations. The only sensible case I can think of that could be hindered are in the presence of double coercions `Coerce_to_int (Coerce_to_pointer expr)`. That can be easilly solved by introducing coercions with a function that checks for that case and rewrite it to `expr`.